src/azul/bigquery_reservation.py

```diff
@@ -0,0 +1,141 @@
+class ResourceNames:
+    commitment: str
+    reservation: str
+    assignment: str
+
+
+class SlotManager(TerraClient):
+    slots = 100
+
+    _reservation_id = 'azul-reindex'
+
+    @cached_property
+    def _client(self) -> ReservationServiceClient:
+        return ReservationServiceClient(credentials=self.credentials)
+
+    @property
+    def _project(self) -> str:
+        return self.credentials.project_id
+
+    @property
+    def _reservation_parent_path(self) -> str:
+        return self._client.common_location_path(
+            project=self._project,
+            location='US'
+        )
+
+    def ensure_slots(self) -> ResourceNames:
+        """
+        Idempotently purchase flex slots.
+        """
+        resources = self.get_resource_names()
+        if resources is None:
+            commitment_name = self._purchase_commitment().name
+            reservation_name = self._create_reservation().name
+            assignment_name = self._create_assignment(reservation_name).name
+            resources = ResourceNames(commitment=commitment_name,
+                                      reservation=reservation_name,
+                                      assignment=assignment_name)
+        else:
+            log.info('Slot commitment already active')
+        return resources
+
+    def delete_slots(self, resources: ResourceNames) -> None:
+        self._client.delete_assignment(name=resources.assignment)
+        self._client.delete_reservation(name=resources.reservation)
+        self._client.delete_capacity_commitment(name=resources.commitment)
+        log.info('Deleted resources: %r', resources)
+
+    def get_resource_names(self) -> Optional[ResourceNames]:
+        args = {}
+        for resource_type, method, parent in [
+            ('commitment', self._client.list_capacity_commitments, self._reservation_parent_path),
+            ('reservation', self._client.list_reservations, self._reservation_parent_path),
+            ('assignment', self._client.list_assignments, self._reservation_parent_path + '/reservations/-')
+        ]:
+            resource_name = self._single_name(resource_type, method(parent=parent))
+            if resource_name is not None:
+                args[resource_type] = resource_name
+        try:
+            return ResourceNames(**args)
+        except TypeError:
+            require(not args, f'BigQuery slot commitment state is inconsistent--'
+                              f'some, but not all resource are missing: {args}')
+            return None
+
+    def _single_name(self, resource_type: str, resources) -> Optional[str]:
+        try:
+            return one((r.name for r in resources),
+                       too_long=RequirementError(f'Too many {resource_type}s in '
+                                                 f'path {self._reservation_parent_path} '
+                                                 f'(should be 0 or 1)'))
+        except ValueError:
```

Comment:
I'm assuming you expect `one` to raise this, but there is too much else going in the body of the try block so this violates https://github.com/DataBiosphere/azul/blob/develop/CONTRIBUTING.rst#120catching-exceptions first bullet.